Apply clone option to target object

creativeux opened the pull request:
closes #208 
closes #186 

* Added ternary for initializing the `destination` object based on the `clone` option.
* Added 2 unit tests... 1 to test that target type info is retained, and another to ensure that target memory pointer is maintained when `clone` is set to `false`.

Commit:
Apply clone option to target object

TehShrike commented:
This looks good.  One request – could you add a test that asserts that when `merge.all` is called with `{ clone: true }`, the first element is mutated/returned?

Commit:
Apply target mutation to merge.all first array entry

creativeux commented:
> could you add a test that asserts that when `merge.all` is called with `{ clone: true }`, the first element is mutated/returned?

Great call, I hadn't used that entrypoint so didn't think to change it. I made the change, but it required a few more lines to change to make it elegant and aligned with the code style in place. The `firstArrayEntry()` method is there to avoid chained ternaries which can be confusing to read, and I moved the logic to determine if the destination should be a new object or a target object into a function.

Let me know if you'd like any add'l changes!

Commit:
Missing index on return

Commit:
Index madness. Oops.

RebeccaStevens commented:
I like it but I am a little worried about backwards compatibility. As is, this would definitely need to be released as a new major version.

creativeux commented:
> I like it but I am a little worried about backwards compatibility. As is, this would definitely need to be released as a new major version.

You're probably right. This is an awkward one because it sounds like it is technically a bug fix because the API "lies" to the consumer about how the clone option is applied, but fixing it potentially breaks consumers whose code relies on the bug being there. I'll let you guys duke it out 😉

TehShrike commented:
Yup, this would be a breaking change.  I'm thinking to pair it with removing the deprecation warning from the `clone` flag

RebeccaStevens commented:
**I've changed my mind on liking this.**

I like the idea behind it but this is definitely a breaking change that I can see a lot of push back happening on if it goes live.

The current implementation of the `clone` option makes it so that if set to `false`, objects will be directly copied over to the resulting merged object.
The new behavior in this MR as is makes the `clone` option when set to `false` will mutate the target option.

These two behaviors are very different and I don't think it would be a good idea to switch from one to the other using the same option name.

The issue with having this behavior on the "clone" flag is that it that it is valid to want to both want to clone and not mutate the any of the parameters.

I would be best to put this behavior behind a new flag called "mergeWithTarget".

I just made a new issue #212 with my proposal for the clone behavior.

Commit:
Add cloneWithTarget flag

Commit:
merge with target, not clone with target...

creativeux commented:
@RebeccaStevens I lied about how long it would take. My wife started watching The Bachelorette and that was my cue to zone out. Let me know if this addresses the concerns re: `clone` vs `mergeWithTarget`.

RebeccaStevens commented on index.js:
```diff
@@ -51,8 +55,17 @@ function propertyIsUnsafe(target, key) {
 
+// Retrieves either a new object or the appropriate target object to mutate.
+function getDestinationObject(target, options) {
+	if (options && !!options.mergeWithTarget) {
```
The `!!` isn't needed.
`if` statements don't care if you pass them a boolean or truthy/falsy value.

creativeux replied:
Good call. Fixed.

RebeccaStevens commented:
Looks good to me :+1:

RebeccaStevens commented:
closes #186

Commit:
Remove unnecessary boolean cast

creativeux commented on index.js:
```diff
@@ -51,8 +55,17 @@ function propertyIsUnsafe(target, key) {
 
+// Retrieves either a new object or the appropriate target object to mutate.
+function getDestinationObject(target, options) {
+	if (options && options.mergeWithTarget) {
```
@RebeccaStevens does this need to account for `options.clone` scenarios for objects that are not the root target? I've read through it a few times and with `deepmerge` being self-referenced, I wonder if there's a case where this function would be called in reference to an object nested deeper than the entrypoint where we might want to return the target here in the case where `options.clone` is `false`?

RebeccaStevens replied:
No I don't think so. This function is only called at most once per call of the main deepmerge function.

creativeux replied:
Per my last commit, this is called repeatedly during recursion down an object tree. Not sure if this changes your thought on the topic.

creativeux replied:
Specifically, `deepmerge` is returned from `cloneUnlessOtherwiseSpecified` which can call `mergeObject` which calls `getDestinationObject`.

RebeccaStevens replied:
I think I'm seeing the issue you are talking about here.

I think with `options.mergeWithTarget: true` & `options.clone: false`, merging with each target going down should happen.
I think with `options.mergeWithTarget: true` & `options.clone: true`, merging should only happen with the top most target.

RebeccaStevens replied:
Looking at the codebase, it doesn't really look set up to be able to add this functionality easily in a non-hacky way.
A bunch of refactoring may be needed (which I'm in the process of doing in #215).

If you want to just do it in a hacky way, you could just remove `mergeWithTarget` from the options object before it is passed to the recursive call.

Commit:
Handle undefined target

creativeux commented:
@TehShrike and @RebeccaStevens - I caught a bug with my in-use fork of this where the merge fails in the case that a target is `undefined`. I pushed a fix and a unit test to ensure that there is always an emtpy object or array to merge into if the target is undefined.

RebeccaStevens commented on index.js:
```diff
@@ -51,8 +55,21 @@ function propertyIsUnsafe(target, key) {
+		return Array.isArray(target) ? firstArrayEntry(target) : target
+	}
+
+	return isArray ? [] : {};
```
The destination object should never be an array.

This is how I would define this function:

```js
function getDestinationObject(target, options) {
	const targetDefined = typeof target !== 'undefined'
	const doMerge = options && (options.mergeWithTarget || options.clone === false)

	if (targetDefined && doMerge) {
		return target
	}

	return {};
}
```

creativeux replied:
For me, the array check was required because during recursion it's possible that there is a property being handled whose value is `undefined`, but we know that the source value is an array, so we need to instantiate an empty object that can be merged into. If I remove this, nested arrays will fail to merge because an array source will try to be merged into an object target.

RebeccaStevens replied:
Yeah sorry, I was thrown off by the function `mergeObject`'s name. I assumed that it's parameters were always objects but in actual fact it merges any 2 values (or returns the latter one if they can't be merged).

RebeccaStevens replied:
Wait no, that's not right. `mergeObject` does only ever receive non-array objects. `target` and `source` should never be arrays in `mergeObject` - arrays should be handled by `options.arrayMerge`.

RebeccaStevens replied:
@creativeux I suspect you'll need to update the logic in `deepmerge` itself to handle the case of `target` being `undefined`.

https://github.com/TehShrike/deepmerge/blob/master/index.js#L87-L93

RebeccaStevens commented on index.js:
```diff
@@ -100,7 +117,7 @@ deepmerge.all = function deepmergeAll(array, options) {
 	return array.reduce(function(prev, next) {
 		return deepmerge(prev, next, options)
-	}, {})
+	}, getDestinationObject(array, undefined, options))
```
Following on from my other comment, this is how `getDestinationObject` would be called from `deepmergeAll`:

```js
const target = array[0];
const destination = getDestinationObject(target, options)
const source = destination === target ? array.slice(1) : array

return source.reduce(function(prev, next) {
	return deepmerge(prev, next, options)
}, destination)
```

Commit:
Improve undefined source / target testing

Commit:
Incomplete assertion definitions
